serving/src/main/java/ai/djl/serving/wlm/WorkLoadManager.java

```diff
@@ -1,119 +1,48 @@
-/*
- * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
- *
- * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
- * with the License. A copy of the License is located at
- *
- * http://aws.amazon.com/apache2.0/
- *
- * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
- * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
- * and limitations under the License.
- */
 package ai.djl.serving.wlm;
 
-import ai.djl.serving.util.ConfigManager;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
+import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
 import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 
-class WorkLoadManager {
+public class WorkLoadManager {
 
-    private ConfigManager configManager;
-    private AtomicInteger gpuCounter;
-    private ExecutorService threadPool;
-    private ConcurrentHashMap<String, List<WorkerThread>> workers;
+    private Executor defaultPredictionExecutor;
+    private Map<String, ExecutorService> specialisedPredictionExecutor;
 
-    public WorkLoadManager(ConfigManager configManager) {
-        this.configManager = configManager;
-        this.gpuCounter = new AtomicInteger(0);
-        threadPool = Executors.newCachedThreadPool();
-        workers = new ConcurrentHashMap<>();
+    public WorkLoadManager() {
+        specialisedPredictionExecutor = new ConcurrentHashMap<>();
+        defaultPredictionExecutor =
```

Comment:
We can't workers share the same thread pool ? Is this because you need to cache the predictor for a thread ?

The auto scale support and the mentioned behaviour was exactly what I had in mind with the build-in java executor solution. I think the java-ThreadPoolExecutor provides exactly this functionality out of the box. You can set a core-pool-size which defines the number of idle worker threads when the system is not under load. When more thread are needed  - cause the work-queue of all existing threads exceeds their capacity - then the ThreadPoolExecutor  automatically scales up to the limit "max-threads". You can set a timeout after which idle Threads automatically scale down to the coreWorkerThread-Limit...if I remember correctly. So I guess we can get the behaviour for free and it's well tested from Oracle. 
![ThreadPoolExecutorService_default_behaviour](https://user-images.githubusercontent.com/16692157/101183428-6b8e5500-3647-11eb-9878-d8516ab1b573.png)

I was not aware that Predictor creation can be heavy, but Executor-Worker-Thread can pull Predictors from a ObjectPool or from a ThreadLocal variable. so for example Threads can be reused for other prediction work. To avoid the overhead of thread-management for idle threads.